csrc/ops.h

```diff
@@ -70,6 +70,13 @@ torch::Tensor awq_gemm(
   torch::Tensor _scaling_factors,
   torch::Tensor _zeros,
   int split_k_iters);
+
+void marlin_gemm(
```

Comment:
May we add the `SMS` parameter for different GPUs? e.g.
```python3
# Pass the SM count for known GPUs to avoid the kernel having to query this information (this is very minor)
gpu = torch.cuda.get_device_name(0)
if 'A100' in gpu:
    SMS = 108
elif 'A10' in gpu:
    SMS = 72
elif '3090' in gpu:
    SMS = 82
elif 'A6000' in gpu:
    SMS = 84
else:
    SMS = -1
```
Besides that, do we really need the `workspace` parameter?
> @workspace: `torch.int` tensor with at least as many entries as there a GPU SMs (256 is usually safe)
Maybe we can use the default value.
```python3
workspace = torch.zeros(256, device=torch.device('cuda:0'))
```

Reply:
When the batch is `16`, set the `thread_k` to 16 and `thread_n` to 256 is ok according to the bench. So may we also add the `thread_k` and `thread_n` parameters?

Reply:
Hi,

I realized that I made a mistake when documenting the `workspace` parameter, it should actually be of size at least `in_features / 256`, I will fix that soon. The main reason why this is an extra parameter is that I found on-the-fly allocation and deallocation during the kernel launch to slow things down a bit, hence I made this an external buffer (note that it can, in principle, be reused between different layers as long as it is large enough).

The kernel should be able to determine the SM count of the GPU automatically, I just fixed it for benchmarks to avoid any minor overheads that are avoidable.

Currently, the kernel only supports `(thread_k, thread_n)` of `(128, 128)` and `(64, 256)`, the former is used for batchsize `<= 16` and the latter otherwise. There are more configurations that could be supported for certain batchsizes, but I found those two to almost always perform best (or very close) and hence decided to simplify it for now.

Reply:
> Hi,
> 
> I realized that I made a mistake when documenting the `workspace` parameter, it should actually be of size at least `in_features / 256`, I will fix that soon. The main reason why this is an extra parameter is that I found on-the-fly allocation and deallocation during the kernel launch to slow things down a bit, hence I made this an external buffer (note that it can, in principle, be reused between different layers as long as it is large enough).
> 
> The kernel should be able to determine the SM count of the GPU automatically, I just fixed it for benchmarks to avoid any minor overheads that are avoidable.
> 
> Currently, the kernel only supports `(thread_k, thread_n)` of `(128, 128)` and `(64, 256)`, the former is used for batchsize `<= 16` and the latter otherwise. There are more configurations that could be supported for certain batchsizes, but I found those two to almost always perform best (or very close) and hence decided to simplify it for now.

Hi @efrantar Thanks for your explanation!